rir/src/compiler/analysis/scope.h

```diff
@@ -18,19 +18,178 @@ namespace pir {
  * all loads (`LdVar`).
  *
  */
-class ScopeAnalysis {
-    std::unordered_set<Instruction*> observedStores;
-    std::unordered_set<Value*> allStoresObserved;
+class ScopeAnalysisState {
+    AbstractREnvironmentHierarchy envs;
+    std::unordered_map<Instruction*, AbstractPirValue> returnValues;
+    AbstractPirValue result;
+    std::set<Instruction*> observedStores;
+    std::set<Value*> allStoresObserved;
+
+    bool mayUseReflection = false;
+
+    AbstractResult mergeCall(Code* cur, const ScopeAnalysisState& other) {
```

Comment:
Not sure why mergeCall does not need to merge the return values, but still, would be good to factorize here.